### ISSUE TITLE
[Feature Request] Support for Inline Suppression of Specific Rule Violations in Apex (e.g., @IsTest(SeeAllData=true))

### ISSUE THREAD

**Issue by FedorAtAdesso** [forcedotcom/code-analyzer#1808]
### Need/Problem Description

In our codebase, we enforce strict rules against certain patterns such as @IsTest(SeeAllData=true) due to best practices. However, there are rare but valid cases, such as tests that call Connect API classes for Commerce Cloud, where SeeAllData=true is necessary.

### Describe your Desired Solution

Currently, there is no apparent way to suppress specific rule violations inline in Apex code. We'd like to request support for inline suppression (e.g., via //NOPMD or similar syntax), so we can selectively override violations in justified cases without disabling the rule globally.

This would allow us to:

- Maintain strict rule enforcement project-wide
- Clearly indicate and document exceptions inline
- Avoid false positives in reports

Please consider adding support for this feature to improve rule flexibility in real-world Apex scenarios.

### Alternatives Considered

The alternative is to not block violations or override the rule globally to lower the score, both isn't desirable

### Urgency

High

**Comment by FedorAtAdesso** [comment]
The example I gave in the description (SeeAllData=true in tests that require it due to technical limitations) is the first one we encountered where. 

Annotations won't be accepted by the salesforce platform or language server and we would have to write a script to strip them before deployment. additionally i was under the impression, since there is no documentation on this that i could find, that inline comments won't work either.

if inline comments can prevent rules from being triggered, it would would be great to have at least a link per engine as to how those need to look like and which rules can not be suppressed at all, if any.

**Comment by stephen-carter-at-sf** [comment]
I'm sorry, this repo is for the Salesforce Code Analyzer.

Are you referring to a rule violation from Salesforce Code Analyzer, or are you referring to something else like an Apex compile error?

By
> In our codebase, we enforce strict rules against certain patterns such as @istest(SeeAllData=true) due to best practices.
how are you enforcing it? With Salesforce Code Analyzer?

What engine / rule are you trying to suppress? I'm refering to what is available with `sf code-analyzer rules -r all --view detail`

**Comment by stephen-carter-at-sf** [comment]
If you are referring to https://pmd.github.io/pmd/pmd_rules_apex_bestpractices.html#apexunittestshouldnotuseseealldatatrue then why wouldn't //NOPMD work for you? Have you tried it?

It would help if you gave me an example file and the command you are trying to run that replicates the rule violation that you wish to suppress.

**Comment by FedorAtAdesso** [comment]
Thank you for reminding me that this repository is for the Salesforce Code Analyzer. I must've briefly forgotten.

As stated previously, we were under the impression that suppression via code comments would not work, since the documentation didn't mention it.

While I feel uncomfortable sharing code (even test code) from our clients, I can provide you with an example of a test class that has the SeeAllData flag set to true in the @IsTest annotation. The default ruleset of the Code Analyzer classifies this as a severity 2 violation.

If it's possible to use different suppression flags for different engines, that could be a viable workaround in the meantime. We appreciate the proposed update to the documentation—it would help clarify which suppression mechanisms apply to which engines, and which rules may not be suppressible at all.

**Comment by tggagne** [comment]
I'm having this same problem.

At the beginning of my class method, after "sf code-analyzer run" complained...
```
 45  3 (Moderate) pmd:CognitiveComplexity                YourClass.cls:6:16            The class 'YourClass' has a total cognitive complexity of 319 (highest 9), current threshold is 50         
 46  3 (Moderate) pmd:CyclomaticComplexity               YourClass.cls:6:16            The class 'YourClass' has a total cyclomatic complexity of 99 (highest 9)
```
I added 
```
@SuppressWarning('PMD.CognitiveComplexity')
@SuppressWarning('PMD.CyclomaticComplexity')
public virtual class YourClass {
``` 
But still got the complaints.

I'd love to be able to disable those rules in the code-analyzer.yml file, but haven't figured out if it can be done there.

**Comment by stephen-carter-at-sf** [comment]
@HazenFley you might be encountering the issue here: https://github.com/pmd/pmd/issues/5935

While you wait for PMD to fix this issue that the introduced, the workaround for you is to remove the spaces so that you have 
```
@SuppressWarnings('PMD.TooManyFields,PMD.ExcessivePublicCount,PMD.FieldNamingConventions')
```
instead of
```
@SuppressWarnings('PMD.TooManyFields, PMD.ExcessivePublicCount, PMD.FieldNamingConventions')
```

**Comment by stephen-carter-at-sf** [comment]
@HazenFley Regarding your comment 
> Additionally, the "quick fix" option [mentioned in the documentation](https://developer.salesforce.com/docs/platform/salesforce-code-analyzer/guide/analyze-vscode.html?q=suppress#suppress-code-analyzer-pmd-violations) does not exist for me.
Can you be more specific as to what you are seeing? If that issue is not related to this issue, then please create a new issue over at: https://github.com/forcedotcom/code-analyzer/issues (click the New Issue button)

**Comment by namrata111f** [comment]
We can now use the suppression markers introduced to filter out violations in particular blocks of code: https://developer.salesforce.com/docs/platform/salesforce-code-analyzer/guide/suppress-violations.html
